### PR TITLE
Fix gem for rails 7

### PR DESCRIPTION
**Pull request by piyushswain**
This PR aims to fix this gem for Rails 7 and latest Solidus.

- Fixed an issue with autoloading constants in initializers
- Fixed `coffee-rails` dependency issue
- Removed `omniauth-amazon` dependency (No longer being maintained)
- Removed `omniauth-twitter` dependecy (No longer being maintained)
- Install Browser tools for CircleCI to fix failing browser tests
- Updated README
- Other fixes for rails 7 compatibility

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -5,7 +5,7 @@ orbs:
   # run specs against Solidus supported versions only without the need
   # to change this configuration every time a Solidus version is released
   # or goes EOL.
-  solidusio_extensions: solidusio/extensions@volatile
+  solidusio_extensions: solidusio/extensions@0.2.27
```
Comment (waiting-for-dev) [inline .circleci/config.yml]: Why do we need that?

File: lib/solidus_social/config.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+require 'solidus_social/social_configuration'
+
+module Spree
+  def self.social_config
+    @config ||= Spree::SocialConfiguration.new
+  end
+
+  def self.const_missing(name)
+    name == :SocialConfig ? social_config : super
```
Comment (waiting-for-dev) [inline lib/solidus_social/config.rb]: Wouldn't it work to append

```ruby
::Spree::SocialConfig = Spree::SocialConfiguration.new
```

in `lib/solidus_social/social_configuration.rb`?

File: lib/solidus_social/social_configuration.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
```
Comment (waiting-for-dev) [inline lib/solidus_social/social_configuration.rb]: I think that we need to remove `app/models/spree/social_configuration.rb` once we move it to `lib`.

File: solidus_social.gemspec [hunk 4]
```diff
@@ -39,6 +39,8 @@ Gem::Specification.new do |spec|
   spec.add_dependency 'solidus_auth_devise'
   spec.add_dependency 'solidus_core', ['>= 2.0.0', '< 4']
   spec.add_dependency 'solidus_support', '~> 0.5'
+  spec.add_dependency 'multi_json'
```
Comment (waiting-for-dev) [inline solidus_social.gemspec]: I can't see any reference to a `MultiJson` constant. Have we tried just removing the dependency?

File: solidus_social.gemspec [hunk 5]
```diff
@@ -39,6 +39,8 @@ Gem::Specification.new do |spec|
...
+  spec.add_dependency 'coffee-rails'
```
Comment (waiting-for-dev) [inline solidus_social.gemspec]: I can't see coffee files. Maybe there were in the past, but it's no longer the case. Does anything break if we remove the dependency?

**Review by waiting-for-dev — CHANGES_REQUESTED** [review]
Thanks, @piyushswain!!

I left some questions.

Also, would you mind fixing CI as part of this PR? If we need to release a new version so that `solidus_bolt` can pull it, we need it to be green.

File: solidus_social.gemspec (hunk 5 above)
Comment (piyushswain) [reply solidus_social.gemspec]: Yes
The install fails and gives the following error message
```
cannot load such file -- coffee_script (LoadError)
```

Comment (waiting-for-dev) [reply solidus_social.gemspec]: Probably we can remove this line:

https://github.com/solidusio-contrib/solidus_social/blob/d97cecd7c1f53093fe37d8f47b3ee41e9bbad7f8/lib/solidus_social/engine.rb#L9

File: solidus_social.gemspec (hunk 4 above)
Comment (piyushswain) [reply solidus_social.gemspec]: The install fails with the following error when `multi_json` is removed
```
cannot load such file -- multi_json (LoadError)
```

Comment (waiting-for-dev) [reply solidus_social.gemspec]: Could you provide the entire stack of the error? :pray:

Comment (piyushswain) [reply solidus_social.gemspec]:
```
Traceback (most recent call last):
	42: from bin/rails:4:in `<main>'
	41: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	40: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	39: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/railties-7.0.3.1/lib/rails/commands.rb:18:in `<main>'
	38: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/railties-7.0.3.1/lib/rails/command.rb:48:in `invoke'
	37: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/railties-7.0.3.1/lib/rails/command/base.rb:87:in `perform'
	36: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/thor-1.2.1/lib/thor.rb:392:in `dispatch'
	35: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/thor-1.2.1/lib/thor/invocation.rb:127:in `invoke_command'
	34: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/thor-1.2.1/lib/thor/command.rb:27:in `run'
	33: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/railties-7.0.3.1/lib/rails/commands/console/console_command.rb:101:in `perform'
	32: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/railties-7.0.3.1/lib/rails/command/actions.rb:14:in `require_application_and_environment!'
	31: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/railties-7.0.3.1/lib/rails/command/actions.rb:22:in `require_application!'
	30: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	29: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	28: from /Users/piyushswain/nebulab/solidus_social_test_app/config/application.rb:7:in `<main>'
	27: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bundler-2.3.6/lib/bundler.rb:176:in `require'
	26: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bundler-2.3.6/lib/bundler/runtime.rb:44:in `require'
	25: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bundler-2.3.6/lib/bundler/runtime.rb:44:in `each'
	24: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bundler-2.3.6/lib/bundler/runtime.rb:55:in `block in require'
	23: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bundler-2.3.6/lib/bundler/runtime.rb:55:in `each'
	22: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bundler-2.3.6/lib/bundler/runtime.rb:60:in `block (2 levels) in require'
	21: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	20: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	19: from /Users/piyushswain/nebulab/solidus_social/lib/solidus_social.rb:8:in `<main>'
	18: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	17: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	16: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	15: from /Users/piyushswain/nebulab/solidus_social/lib/solidus_social/engine.rb:7:in `<main>'
	14: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	13: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	12: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	11: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/omniauth-amazon-1.0.1/lib/omniauth-amazon.rb:1:in `<main>'
	10: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	 9: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	 8: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	 7: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/omniauth-amazon-1.0.1/lib/omniauth/amazon.rb:2:in `<main>'
	 6: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	 5: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	 4: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	 3: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/omniauth-amazon-1.0.1/lib/omniauth/strategies/amazon.rb:2:in `<main>'
	 2: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
	 1: from /Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:15:in `require'
/Users/piyushswain/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:15:in `require': cannot load such file -- multi_json (LoadError)
```

File: solidus_social.gemspec (hunk 5 above)
Comment (piyushswain) [reply solidus_social.gemspec]: Yeah, this can probably be removed

File: .circleci/config.yml (hunk 1 above)
Comment (piyushswain) [reply .circleci/config.yml]: Removed the commit that changed this

File: lib/solidus_social/social_configuration.rb (hunk 3 above)
Comment (piyushswain) [reply lib/solidus_social/social_configuration.rb]: Removed this file here [6d93219](https://github.com/solidusio-contrib/solidus_social/commit/6d9321941a614f0dd7541e5e00060a02c01dc75b)

File: solidus_social.gemspec (hunk 5 above)
Comment (piyushswain) [reply solidus_social.gemspec]: Removed and fixed the dependency issue here [382d2b](https://github.com/solidusio-contrib/solidus_social/commit/382d2b15f9afb1843bccffb40dbe983539083118)

File: lib/solidus_social/config.rb (hunk 2 above)
Comment (piyushswain) [reply lib/solidus_social/config.rb]: Modified this here [4b71338](https://github.com/solidusio-contrib/solidus_social/commit/4b71338044bcb42b46f8e8db2064cf33bae63920)

File: solidus_social.gemspec (hunk 4 above)
Comment (waiting-for-dev) [reply solidus_social.gemspec]: I see. It's a problem with amazon's Omniauth integration. I created a [PR](https://github.com/wingrunr21/omniauth-amazon/pull/19), but it doesn't look to be very well maintained.

File: lib/solidus_social/config.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+require 'solidus_social/social_configuration'
+
+module Spree
```
Comment (waiting-for-dev) [inline lib/solidus_social/config.rb]: No need to wrap within the `Spree` module. I'd just add the `::Spree::SocialConfig` assignment in the `social_configuration.rb` file.

**Comment by waiting-for-dev** [comment]
Ref: https://github.com/solidusio-contrib/solidus_social/pull/102#discussion_r938726743

I'd remove both [omniauth-amazon](https://github.com/wingrunr21/omniauth-amazon) & [omniauth-twitter](https://github.com/arunagw/omniauth-twitter) from solidus_social, as they are the only integration that doesn't look like maintained anymore. Thoughts @aldesantis, @kennyadsl?

**Comment by kennyadsl** [comment]
@waiting-for-dev that makes sense, especially if they are blocking us from upgrading this extension.

**Comment by waiting-for-dev** [comment]
> @waiting-for-dev that makes sense, especially if they are blocking us from upgrading this extension.

Yeah, we just need to make sure we release a new major. @piyushswain, we can do it as part of this PR, if you like.

**Comment by piyushswain** [comment]
> > @waiting-for-dev that makes sense, especially if they are blocking us from upgrading this extension.
> 
> Yeah, we just need to make sure we release a new major. @piyushswain, we can do it as part of this PR, if you like.

Okay, so I will remove `omniauth-twitter` and `omniauth-amazon` from the gem so that we can avoid the `multi-json` error.

File: lib/solidus_social/config.rb (hunk 6 above)
Comment (piyushswain) [reply lib/solidus_social/config.rb]: Modified this here [36f081b](https://github.com/solidusio-contrib/solidus_social/commit/36f081b55005ad895a564c8c2cf34e63cb63b325)

Fixed now

**Comment by waiting-for-dev** [comment]
@piyushswain, it looks like tests are failing

**Comment by piyushswain** [comment]
@waiting-for-dev The tests pass on my development environment but a few tests fail on the CI with the following error
```
Webdrivers::BrowserNotFound:
            Failed to find Chrome binary.
```
All the failing tests have the same error.

I'm not sure if this is an issue in the CI or something to do with the gem itself.

**Comment by waiting-for-dev** [comment]
> @waiting-for-dev The tests pass on my development environment but a few tests fail on the CI with the following error
> 
> ```
> Webdrivers::BrowserNotFound:
>             Failed to find Chrome binary.
> ```
> 
> All the failing tests have the same error.
> 
> I'm not sure if this is an issue in the CI or something to do with the gem itself.

@piyushswain, CircleCI images changed and now you need to require the browser tools manually. I think you only need to do the same that this commit:

https://github.com/solidusio/solidus_auth_devise/commit/e2503aa5b83c2448d686b3d7e8e7a87e1628877a

You can use [latest v1.4](https://circleci.com/developer/orbs/orb/circleci/browser-tools), though.

**Comment by waiting-for-dev** [comment]
Cool! We've fixed the CircleCI issue. There're still some failures. Which version of Rails did you use locally for testing? If it was minor than 7, probably there's more to fix to make it compatible with Rails 7.

**Comment by piyushswain** [comment]
> Cool! We've fixed the CircleCI issue. There're still some failures. Which version of Rails did you use locally for testing? If it was minor than 7, probably there's more to fix to make it compatible with Rails 7.

Tests run on the Solidus master branch on my local environment, so Rails 7
I am currently trying to fix the issues which are showing up on the CI

**Comment by piyushswain** [comment]
> Cool! We've fixed the CircleCI issue. There're still some failures. Which version of Rails did you use locally for testing? If it was minor than 7, probably there's more to fix to make it compatible with Rails 7.

Hey @waiting-for-dev, I just noticed, the `sign_in_specs` related to facebook have been failing since today morning even on the `master` branch for solidus which were previously passing on local as well.
Investigating on the issue now.

File: .circleci/config.yml [hunk 7]
```diff
@@ -1,6 +1,7 @@
 version: 2.1
 
 orbs:
+  browser-tools: circleci/browser-tools@1.1
```
Comment (waiting-for-dev) [inline .circleci/config.yml]:
```suggestion
  browser-tools: circleci/browser-tools@1.4
```

**Comment by waiting-for-dev** [comment]
Thanks for your continuous work on it, @piyushswain. We're almost there! I left a suggestion, as we can update the browser tools versions.

Given that `omniauth-facebook` is forcing us to downgrade the support for `omniauth`, I'd also remove that provider. As we already support adding any provider manually, users can still easily do it by themselves. In fact, I think we should probably avoid adding any integration OOTB, as we're adding dependencies that some users won't need. But that's out of the scope of this PR.

Finally, we also need to update the README with the changes.

**Comment by piyushswain** [comment]
> Thanks for your continuous work on it, @piyushswain. We're almost there! I left a suggestion, as we can update the browser tools versions.
> 
> Given that `omniauth-facebook` is forcing us to downgrade the support for `omniauth`, I'd also remove that provider. As we already support adding any provider manually, users can still easily do it by themselves. In fact, I think we should probably avoid adding any integration OOTB, as we're adding dependencies that some users won't need. But that's out of the scope of this PR.
> 
> Finally, we also need to update the README with the changes.

It's not that `omniauth-facebook` is forcing us to downgrade the `omniauth` version. The `omniauth` version was 1.9.1 before this PR as well.
`omniauth-facebook` isn't allowing us to update the `omniauth` version though.

But yes, we could remove it in another PR and users can add it later by themselves.

File: .circleci/config.yml (hunk 7 above)
Comment (piyushswain) [reply .circleci/config.yml]: Fixed here [d9fc568](https://github.com/solidusio-contrib/solidus_social/commit/d9fc5683dd3ca6ce0423e31baf2250a6f72fa1b1)

**Comment by waiting-for-dev** [comment]
> It's not that `omniauth-facebook` is forcing us to downgrade the `omniauth` version. The `omniauth` version was 1.9.1 before this PR as well. `omniauth-facebook` isn't allowing us to update the `omniauth` version though.
> 
> But yes, we could remove it in another PR and users can add it later by themselves.

Well, yeah, but I mean that it was open in the gemspec, and now we're requiring it to be lesser than v2. I'll definitely remove facebook instead of forcing users to use old versions of omniauth. It's not that they need to create a PR here, they can add facebook only in their apps if they need it. See https://github.com/solidusio-contrib/solidus_social#other-oauth-providers

**Review by DanielePalombo — APPROVED** [review]
LGTM.

Thank you @piyushswain. 
I have opened an issue to [remove omniauth-facebook](https://github.com/solidusio-contrib/solidus_social/issues/104).

**Comment by DanielePalombo** [comment]
> > It's not that `omniauth-facebook` is forcing us to downgrade the `omniauth` version. The `omniauth` version was 1.9.1 before this PR as well. `omniauth-facebook` isn't allowing us to update the `omniauth` version though.
> > But yes, we could remove it in another PR and users can add it later by themselves.
> 
> Well, yeah, but I mean that it was open in the gemspec, and now we're requiring it to be lesser than v2. I'll definitely remove facebook instead of forcing users to use old versions of omniauth. It's not that they need to create a PR here, they can add facebook only in their apps if they need it. See https://github.com/solidusio-contrib/solidus_social#other-oauth-providers

@waiting-for-dev I have opened an [issue](https://github.com/solidusio-contrib/solidus_social/issues/104) to remove omniauth-facebook and the commit that locks the version. Is there something else to do with this PR? Otherwise, can you review it again so we can merge it? 🙏 

Thank you for your support.

**Comment by waiting-for-dev** [comment]
Hey @DanielePalombo. No, we can merge it. But it'd be nice to remove `omniauth-facebook` before releasing the new version of it.

**Review by waiting-for-dev — APPROVED** [review]

**Pull request merged**